nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-file-authorizer/src/main/java/org/apache/nifi/authorization/FileUserGroupProvider.java

```diff
@@ -455,36 +466,92 @@ public void setNiFiProperties(NiFiProperties properties) {
     @Override
     public synchronized void inheritFingerprint(String fingerprint) throws AuthorizationAccessException {
         final UsersAndGroups usersAndGroups = parseUsersAndGroups(fingerprint);
-        usersAndGroups.getUsers().forEach(user -> addUser(user));
-        usersAndGroups.getGroups().forEach(group -> addGroup(group));
+        inherit(usersAndGroups);
+    }
+
+    private synchronized void inherit(final UsersAndGroups usersAndGroups) {
+        addUsersAndGroups(usersAndGroups);
+    }
+
+    @Override
+    public synchronized void forciblyInheritFingerprint(final String fingerprint) throws AuthorizationAccessException {
+        if (fingerprint == null || fingerprint.trim().isEmpty()) {
+            logger.info("Inheriting Empty Users & Groups. Will backup existing Uesrs & Groups first.");
+            backupUsersAndGroups();
+            purgeUsersAndGroups();
+
+            return;
+        }
+
+        final UsersAndGroups usersAndGroups = parseUsersAndGroups(fingerprint);
+
+        if (isInheritable(usersAndGroups)) {
+            logger.debug("Inheriting cluster's Users & Groups");
+            inherit(usersAndGroups);
+        } else {
+            logger.info("Cannot directly inherit proposed Users & Groups so will backup existing Users & Groups and then replace with proposed configuration");
+            backupUsersAndGroups();
+            purgeUsersAndGroups();
+            addUsersAndGroups(usersAndGroups);
+        }
+    }
+
+    public void backupUsersAndGroups() throws AuthorizationAccessException {
```

Comment:
Should `backupUsersAndGroups` and `purgeUsersAndGroups` be a requirement for all `ConfigurableUserGroupProviders`? It contains the corresponding methods for inheritance. I guess the question is whether all `ConfigurableUserGroupProviders` should support backing up and purging or whether this is an implementation detail of the file-based implementations. Based off the JavaDocs it sounds like implementations should do this if possible as part of the forcible inheritance... so maybe I just answered my own questions.